src/main/kotlin/xyz/avalonxr/command/WelcomeCommand.kt

```diff
@@ -0,0 +1,160 @@
+import xyz.avalonxr.data.CommandResult
+import xyz.avalonxr.data.OptionStore
+import xyz.avalonxr.data.command.CommandBinding
+import xyz.avalonxr.data.entity.WelcomeMessage
+import xyz.avalonxr.data.error.AvalonError
+import xyz.avalonxr.data.error.CommandError
+import xyz.avalonxr.dsl.discord.OptionType
+import xyz.avalonxr.dsl.discord.command
+import xyz.avalonxr.dsl.discord.option
+import xyz.avalonxr.dsl.discord.optional
+import xyz.avalonxr.dsl.discord.subCommand
+import xyz.avalonxr.enums.CommandScope
+import xyz.avalonxr.models.discord.Command
+import xyz.avalonxr.repository.WelcomeMessageRepository
+import kotlin.jvm.optionals.getOrDefault
+import kotlin.jvm.optionals.getOrNull
+
+@AvalonCommand
+class WelcomeCommand @Autowired constructor(
+    private val messageRepository: WelcomeMessageRepository,
+) : Command {
+
+    override fun getBinding(): CommandBinding = CommandBinding(
+        command("welcome") {
+            description("Welcome command")
+
+            subCommand {
+                name("set")
+                description("Configure welcome message")
+
+                option(OptionType.STRING) {
+                    name("message")
+                    description("Welcome Message")
+                }
+
+                option(OptionType.CHANNEL) {
+                    name("channel")
+                    description("Welcome Channel")
+                    optional()
+                }
+            }
+
+            subCommand {
+                name("preview")
+                description("Preview welcome message")
+
+                option(OptionType.BOOLEAN) {
+                    name("formatted")
+                    description("Preview formatted or un-formatted message")
+                    optional()
+                }
+            }
+        },
+        CommandScope.GUILD
+    )
+
+    override fun processCommand(
+        source: ChatInputInteractionEvent,
+        options: OptionStore
+    ): CommandResult {
+        return when (options.subCommand) {
+            "set" -> setSubcommand(source, options)
+            "preview" -> previewSubcommand(source, options)
+            else -> welcomeMessageError(CommandError.CommandNotFound("[set, preview]"), source)
+        }
+    }
+
+    private fun setSubcommand(
+        source: ChatInputInteractionEvent,
+        options: OptionStore
+    ): CommandResult {
+        val guildId = source.interaction.guildId
+            .map(Snowflake::asLong)
+            .getOrNull()
+            ?: return welcomeMessageError(CommandError.CommandNotFromGuild, source)
+        val defaultChannel = source.interaction.channel
+            .block()
+        val channel = when ("channel" in options) {
+            true ->
+                options
+                    .findChannel<TextChannel>("channel")
+                    ?: return welcomeMessageError(CommandError.IncorrectChannelType, source)
+            else -> defaultChannel
+        }
```

Comment:
Wonder if some kind of box type or additional helper might help with the differing cases here, seeing as we don't have an easy way to avoid boilerplate like this currently.